src/algebraic_geometry/AffineScheme.lean

```diff
@@ -307,6 +307,124 @@ begin
+hU.from_Spec.1.c.app (op $ X.basic_open f) ≫ (Scheme.Spec.obj $ op $ X.presheaf.obj (op U))
+  .presheaf.map (eq_to_hom $ (hU.from_Spec_preimage_basic_open f).symm).op
+
+instance {X : Scheme} {U : opens X.carrier} (hU : is_affine_open U)
+  (f : X.presheaf.obj (op U)) : is_iso (basic_open_sections_to_affine hU f) :=
+begin
+  delta basic_open_sections_to_affine,
+  apply_with is_iso.comp_is_iso { instances := ff },
+  { apply PresheafedSpace.is_open_immersion.is_iso_of_subset,
+    rw hU.from_Spec_range,
+    exact RingedSpace.basic_open_subset _ _ },
+  apply_instance
+end
+.
+lemma is_localization_basic_open {X : Scheme} {U : opens X.carrier} (hU : is_affine_open U)
+  (f : X.presheaf.obj (op U)) :
+  is_localization.away f (X.presheaf.obj (op $ X.basic_open f)) :=
+begin
+  apply (is_localization.is_localization_iff_of_ring_equiv (submonoid.powers f)
+    (as_iso $ basic_open_sections_to_affine hU f ≫ (Scheme.Spec.obj _).presheaf.map
+      (eq_to_hom (basic_open_eq_of_affine _).symm).op).CommRing_iso_to_ring_equiv).mpr,
+  convert structure_sheaf.is_localization.to_basic_open _ f,
+  change _ ≫ (basic_open_sections_to_affine hU f ≫ _) = _,
+  delta basic_open_sections_to_affine,
+  erw ring_hom.algebra_map_to_algebra,
+  simp only [Scheme.comp_val_c_app, category.assoc],
+  erw hU.from_Spec.val.c.naturality_assoc,
+  rw hU.from_Spec_app_eq,
+  dsimp,
+  simp only [category.assoc, ← functor.map_comp, ← op_comp],
+  apply structure_sheaf.to_open_res,
+end
+
+lemma basic_open_basic_open_is_basic_open {X : Scheme} {U : opens X.carrier}
+  (hU : is_affine_open U) (f : X.presheaf.obj (op U)) (g : X.presheaf.obj (op $ X.basic_open f)) :
+    ∃ f' : X.presheaf.obj (op U), X.basic_open f' = X.basic_open g :=
+begin
+  haveI := is_localization_basic_open hU f,
+  obtain ⟨x, ⟨_, n, rfl⟩, rfl⟩ := is_localization.surj' (submonoid.powers f) g,
+  use f * x,
+  rw [algebra.smul_def, Scheme.basic_open_mul, Scheme.basic_open_mul],
+  erw Scheme.basic_open_res,
+  refine (inf_eq_left.mpr _).symm,
+  convert inf_le_left using 1,
+  apply Scheme.basic_open_of_is_unit,
+  apply submonoid.left_inv_le_is_unit _ (is_localization.to_inv_submonoid (submonoid.powers f)
+    (X.presheaf.obj (op $ X.basic_open f)) _).prop
+end
+
+lemma exists_basic_open_subset_affine_inter {X : Scheme} {U V : opens X.carrier}
+  (hU : is_affine_open U) (hV : is_affine_open V) (x : X.carrier) (hx : x ∈ U ∩ V) :
+  ∃ (f : X.presheaf.obj $ op U) (g : X.presheaf.obj $ op V),
+    X.basic_open f = X.basic_open g ∧ x ∈ X.basic_open f :=
+begin
+  obtain ⟨f, hf₁, hf₂⟩ := hU.exists_basic_open_subset ⟨x, hx.2⟩ hx.1,
+  obtain ⟨g, hg₁, hg₂⟩ := hV.exists_basic_open_subset ⟨x, hf₂⟩ hx.2,
+  obtain ⟨f', hf'⟩ := basic_open_basic_open_is_basic_open hU f
+    (X.presheaf.map (hom_of_le hf₁ : _ ⟶ V).op g),
+  replace hf' := (hf'.trans (RingedSpace.basic_open_res _ _ _)).trans (inf_eq_right.mpr hg₁),
+  exact ⟨f', g, hf', hf'.symm ▸ hg₂⟩
+
```

Comment:
What happened here? Did half of the proof get lost?

Reply:
oops I accidentally removed the `end`.